controllers/bounce_processes.go

```diff
@@ -237,7 +220,7 @@ func getProcessesReadyForRestart(logger logr.Logger, cluster *fdbv1beta2.Foundat
 	// block the restart command if a process is missing longer than the specified GetIgnoreMissingProcessesSeconds.
 	// Those checks should ensure we only run the restart command if all processes that have to be restarted and are connected
 	// to cluster are ready to be restarted.
-	if cluster.IsBeingUpgraded() && (counts.Total()-missingProcesses-upgradedProcesses) != len(addresses) {
+	if cluster.IsBeingUpgradedWithVersionIncompatibleVersion() && (counts.Total()-missingProcesses) != len(addresses) {
```

Comment:
Why "upgradeProcesses" is not getting included here? Shouldn't this check/logic be same (as before) for version incompatible upgrades?

Reply:
For version incompatible upgrades we have no way to differentiate the case of a process being missing and already being upgraded since those processes will be missing from the cluster status (in theory they should appear in the incompatible processes list). So the `upgradedProcesses` was only interesting during a patch upgrade, with the removal of the staging phase the cluster wide restart was removed.

Reply:
So in the case of version incompatible upgrades if there are incompatible processes (processes at the older version) in the cluster then we will still go ahead with the upgrade, which appears to me as a change in the behavior/logic of version incompatible upgrades. Probably we should have a document that talks about the difference cases that the processes could be in (missing vs all processes present, if missing then incompatible or not, if incompatible then at an older version or at the upgrade version, version compatible/incompatible upgrade or not, what if those processes have a special role (like coordinator), etc., and have a description of how the operator is supposed to handle those cases - to have confidence that the operator can correctly handle those scenarios, and also that would make code reviews easier (and we won't need to have design discussions in PRs).

Reply:
I started the document here: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1511

> So in the case of version incompatible upgrades if there are incompatible processes (processes at the older version) in the cluster then we will still go ahead with the upgrade, which appears to me as a change in the behavior/logic of version incompatible upgrades. 

Can you add some more details on this? I'm not sure I understand that concern based on the code changes. The changes to the code only affected version compatible upgrades.

Reply:
Thanks for starting the documentation!

> The changes to the code only affected version compatible upgrades.

Not totally correct, right? The change above affects version incompatible upgrades - the check used to be "(counts.Total()-missingProcesses-upgradedProcesses) != len(addresses)", now it is "(counts.Total()-missingProcesses) != len(addresses)", that is, you are not subtracting "upgradedProcesses" now?

> For version incompatible upgrades we have no way to differentiate the case of a process being missing and already being upgraded since those processes will be missing from the cluster status

Do we know for sure this is what "status" will show for upgraded processes? As I said in another place(s), I don't think we have server side testing for status output of a cluster that has processes in different versions. We need to write a kubernetes test for it, and verify what it outputs.

So the change you are making here in the context of version incompatible upgrades may be a good change - hopefully we will have an explanation of how missing/upgraded processes will be handled in the case of version incompatible upgrades.